### PR TITLE
[Bugfix][Spec Decode][Structured Output] Drive grammar masks from GPU logit counts

### PR DESCRIPTION
**Pull request by LucasWilkinson**
## Purpose

Fix the DSpark adaptive-verification zero-budget crash with structured outputs by making the grammar-mask application depend on the GPU's finalized per-request logit counts.

The scheduler generates masks from scheduled drafts, but adaptive verification can compact the active logits to bonus-only on the GPU. The existing worker builds one CPU mapping entry per CPU logit row, so the zero-budget compaction makes that mapping shorter than the generated mask and trips `assert num_masks == len(mapping)`.

This change gives every structured-output request a fixed-width mask block. The MRV2 Triton kernel launches per grammar request, reads its actual width from device-side `cu_num_logits`, and applies only those active positions. It no longer needs CPU `cu_num_logits_np`, scheduled-draft counts, bonus counts, or new metadata on `InputBatch`/`GrammarOutput`.

MRV1 shares the mask producer but uses packed logits. Its existing repacking path now indexes each request's fixed-width block before applying xgrammar.

### Relationship to #52436

This is a materially different replacement for #52436, not an independent duplicate: #52436 makes the CPU mapping follow scheduled draft counts, whereas this version removes the per-mask CPU mapping and makes the GPU's finalized counts authoritative. It carries forward the original root-cause work, and #52436's author `oops-oom` is credited as a co-author.

Duplicate searches for `52436 in:body` and `adaptive verification structured output grammar bitmask` found no other implementation of this GPU-driven design. There is no linked issue on #52436 in which to record the design difference.

## Test Plan

```bash
.venv/bin/python -m pytest \
  tests/v1/spec_decode/test_adaptive_verification.py \
  tests/v1/spec_decode/test_mtp_structured_output.py -q

.venv/bin/python -m pytest \
  tests/v1/structured_output/test_backend_guidance.py -q

.venv/bin/python -m pytest \
  tests/v1/worker/test_gpu_rejection_sampler_chunking.py \
  tests/v1/worker/test_gpu_batch_ordering.py -q

.venv/bin/pre-commit run --files \
  tests/v1/spec_decode/test_mtp_structured_output.py \
  vllm/v1/structured_output/__init__.py \
  vllm/v1/structured_output/utils.py \
  vllm/v1/worker/gpu/structured_outputs.py \
  vllm/v1/worker/gpu/warmup.py
```

MRV1 model smoke:

```bash
VLLM_USE_V2_MODEL_RUNNER=0 CUDA_VISIBLE_DEVICES=0 \
  .venv/bin/python .venv/mrv1_structured_spec_smoke.py
```

The local smoke used OPT-125M, prompt-lookup ngram speculation with five speculative tokens, xgrammar, and two simultaneous JSON-schema requests.

A direct MRV2 Triton smoke used two requests with stride four and GPU counts `[1, 1]`. It verified that mask rows 0 and 4 were applied while an inactive mask row was ignored.

## Test Result

```text
adaptive verification + MTP structured output: 21 passed
backend guidance:                           6 passed
rejection sampler chunking + batch order: 15 passed
pre-commit:                         all hooks passed
```

MRV1 initialized `vllm/v1/worker/gpu_model_runner.py`, ran speculative structured generation, and returned valid outputs:

```text
{"answer": 2}
{"answer": 3}
```

The same fixed-stride implementation was also exercised with a full DeepSeek-V4-Flash-0731 TP=4 run, adaptive verification forced to a zero draft budget, and two simultaneous JSON-schema requests. It completed with the same valid outputs instead of the mask/mapping assertion.

## AI assistance

AI assistance was used to analyze the bug, implement the alternative, audit MRV1/MRV2, and run the tests above. This remains a draft pending human review of every changed line before it is marked ready.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. No documentation change is needed for this internal correctness fix.
</details>

### CODE REVIEW

**Comment by oops-oom** [comment]
Agreed on by-req — the control flow is much clearer. I'm curious whether the old by-token mapping was meant as a copy optimization? My read is it was matching the packed variable-length mask/logit layout (one mask row ↔ one logit row) rather than saving on the copy — if anything, by-req copies less (one entry per request vs one per logit row). That 1:1 pairing is also exactly what compaction breaks.

The extra bytes, if any, come from padding short requests to fixed decode_query_len blocks, not from by-req indexing itself — a no-op on a uniform spec batch. Mixed scheduled draft lengths pay more bitmask than the packed layout, but K is small so it should be fine.

One nit: the inner for position_idx in range(MASK_STRIDE) serializes a dimension that used to live in the grid. For small stride (2–6) it's minor, but we could keep it parallel by launching grid=(num_grammar_reqs * MASK_STRIDE, ...) and recovering grammar_idx = pid // STRIDE, position_idx = pid % STRIDE — req_idx/width still come from req_indices + device cu_num_logits, and inactive positions are just cheap masked-off programs like before.

**Comment by sfeng33** [comment]
The PR should fix https://github.com/vllm-project/vllm/issues/52620 as well

**Comment by njhill** [comment]
And should fix https://github.com/vllm-project/vllm/issues/49694 hopefully
